File: libs/computer/typescript/src/interface/factory.ts
import type { BaseComputerInterface } from './base';
import type { OSType } from '../computer';

export interface InterfaceOptions {
  ipAddress: string;
  apiKey?: string;
  vmName?: string;
}

/**
 * Factory for creating OS-specific computer interfaces.
 */
export class InterfaceFactory {
  /**
   * Create an interface for the specified OS.
   * 
   * @param os The operating system type ('macos', 'linux', 'windows')
   * @param ipAddress The IP address to connect to
   * @param apiKey Optional API key for cloud providers
   * @param vmName Optional VM name for cloud providers
   * @returns An instance of the appropriate computer interface
   */
  static createInterfaceForOS(
    os: OSType,
    ipAddress: string,
    apiKey?: string,
    vmName?: string
  ): BaseComputerInterface {
    const options: InterfaceOptions = {
      ipAddress,
      apiKey,
      vmName
    };

    switch (os) {
      case 'macos':
        // Dynamic import would be used in real implementation
        // TODO: Implement macOS interface
        throw new Error('macOS interface not yet implemented');
        
      case 'linux':
        // Dynamic import would be used in real implementation
        // TODO: Implement Linux interface
        throw new Error('Linux interface not yet implemented');
        
      case 'windows':
        // Dynamic import would be used in real implementation
        // TODO: Implement Windows interface
        throw new Error('Windows interface not yet implemented');
        
      default:
        // TODO: Implement interface for this OS
        throw new Error(`Interface for OS ${os} not implemented`);
    }
  }
}

File: libs/python/computer/computer/interface/factory.py
"""Factory for creating computer interfaces."""

from typing import Literal, Optional
from .base import BaseComputerInterface
from .tracing_interface import ITracingManager

class InterfaceFactory:
    """Factory for creating OS-specific computer interfaces."""
    
    @staticmethod
    def create_interface_for_os(
        os: Literal['macos', 'linux', 'windows'],
        ip_address: str,
        api_key: Optional[str] = None,
        vm_name: Optional[str] = None,
        tracing: Optional[ITracingManager] = None
    ) -> BaseComputerInterface:
        """Create an interface for the specified OS.
        
        Args:
            os: Operating system type ('macos', 'linux', or 'windows')
            ip_address: IP address of the computer to control
            api_key: Optional API key for cloud authentication
            vm_name: Optional VM name for cloud authentication
            tracing: Optional tracing manager for logging events
            
        Returns:
            BaseComputerInterface: The appropriate interface for the OS
            
        Raises:
            ValueError: If the OS type is not supported
        """
        # Import implementations here to avoid circular imports
        from .macos import MacOSComputerInterface
        from .linux import LinuxComputerInterface
        from .windows import WindowsComputerInterface
        
        if os == 'macos':
            return MacOSComputerInterface(ip_address, api_key=api_key, vm_name=vm_name, tracing=tracing)
        elif os == 'linux':
            return LinuxComputerInterface(ip_address, api_key=api_key, vm_name=vm_name, tracing=tracing)
        elif os == 'windows':
            return WindowsComputerInterface(ip_address, api_key=api_key, vm_name=vm_name, tracing=tracing)
        else:
            raise ValueError(f"Unsupported OS type: {os}")
